core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1108,11 +1118,14 @@ private[spark] class BlockManager(
   def removeBlock(blockId: BlockId, tellMaster: Boolean = true): Unit = {
     logDebug(s"Removing block $blockId")
     val info = blockInfo.get(blockId).orNull
-    if (info != null) {
+    if (info != null && !pendingToRemove.containsKey(blockId)) {
+      pendingToRemove.put(blockId, currentTaskAttemptId)
       info.synchronized {
+        val level = info.level
         // Removals are idempotent in disk store and memory store. At worst, we get a warning.
-        val removedFromMemory = memoryStore.remove(blockId)
-        val removedFromDisk = diskStore.remove(blockId)
+        val removedFromMemory = if (level.useMemory) memoryStore.remove(blockId) else false
+        pendingToRemove.remove(blockId)
```

Comment:
we should wrap this in a try finally in case an exception is thrown. Otherwise we may never release the "pending" lock:
```
if (info != null && !pendingToRemove.containsKey(blockId)) {
  pendingToRemove.put(blockId, currentTaskAttemptId)
  try {
    info.synchronized {
      ... // remove the block from memoryStore
    }
  } finally {
    pendingToRemove.remove(blockId)
  }
}
```
Here and a few other places.